### ISSUE TITLE
CRIUSupport: Add multi-threaded hooks

### ISSUE THREAD

**Issue by tajila** [eclipse-openj9/openj9#18058]
```
	enum HookMode {
		SINGLE_THREAD,
		MULTI_THREAD
	}
```

```
	/**
	 * User hook that is run before checkpointing the JVM.
	 *
	 * If single-threaded mode is requested, no other application threads
	 * will be active. Users should avoid synchronization of objects that are not owned
	 * by the thread, terminally blocking operations and launching new threads in the hook.
	 * If the thread attempts to acquire a lock that it doesn't own, an exception will
	 * be thrown.
	 *
	 * If multi-threaded mode is requested, the hook will be run alongside all other active Java threads.
	 * 
	 * High priority hooks are run last before checkpoint, and vice-versa for low priority hooks.
	 * The priority of the hook is with respect to the other hooks run within that mode. Multi-threaded
	 * hooks are implicitly lower priority than single-threaded hooks. Ie. the lowest priority single-threaded
	 * hook is higher priority than the highest priority multi-threaded hook. 
	 * 
	 * @param hook user hook
	 * @param hookMode mode in which the hook is run, either multi-threaded or single-threaded
	 * @param priority the priority of the hook, between 0 - 99. Throws UnsupportedOperationException otherwise.
	 * @return this
	 */
	public CRIUSupport registerPreCheckpointHook(Runnable hook, HookMode mode, int priority) throws UnsupportedOperationException
```

```
	/**
	 * User hook that is run after restoring a checkpoint image.
	 *
	 * If single-threaded mode is requested, no other application threads
	 * will be active. Users should avoid synchronization of objects that are not owned
	 * by the thread, terminally blocking operations and launching new threads in the hook.
	 * If the thread attempts to acquire a lock that it doesn't own, an exception will
	 * be thrown.
	 *
	 * If multi-threaded mode is requested, the hook will be run alongside all other active Java threads.
	 * 
	 * High priority hooks are run first after restore, and vice-versa for low priority hooks.
	 * The priority of the hook is with respect to the other hooks run within that mode. Multi-threaded
	 * hooks are implicitly lower priority than single-threaded hooks. Ie. the lowest priority single-threaded
	 * hook is higher priority than the highest priority multi-threaded hook. 
	 * 
	 * @param hook user hook
	 * @param hookMode mode in which the hook is run, either multi-threaded or single-threaded
	 * @param priority the priority of the hook, between 0 - 99. Throws UnsupportedOperationException otherwise.
	 * @return this
	 */
	public CRIUSupport registerPostRestoreHook(Runnable hook, HookMode mode, int priority) throws UnsupportedOperationException
```

**Comment by tajila** [comment]
@JasonFengJ9 Please take a look at this

**Comment by tajila** [comment]
The multi-threaded hooks should be called in `checkpointJVM` but outside of `checkpointJVMImpl`.

We can bump up the priorities for below to make room for the user priorities

```
	private static final int RESTORE_CLEAR_INETADDRESS_CACHE_PRIORITY = 100;
	private static final int RESTORE_ENVIRONMENT_VARIABLES_PRIORITY = 100;
	private static final int RESTORE_SYSTEM_PROPERTIES_PRIORITY = 100;
```

**Comment by tajila** [comment]
We need to add two new lists to `J9InternalCheckpointHookAPI` for the multi-threaded variants.

**Comment by tajila** [comment]
The existing APIs can redirect to the new ones with `USER_HOOKS_PRIORITY` and `HookMode.SINGLE_THREAD` as defaults.

**Comment by JasonFengJ9** [comment]
Will look into this.

**Comment by tajila** [comment]
> For multi-thread hooks, they might require Thread.join() before/after invoking checkpointJVMImpl().

Multithread doesnt necesarrily mean run in parallel; I guess concurrent is a better definition. It simply means we don't run them in single thread mode. In any case, they would need to be run serially in order to respect the priorities.
